fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/sorting/MemorySorter.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * TreeMapSorter.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.sorting;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.RecordCoreArgumentException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorContinuation;
+import com.apple.foundationdb.record.provider.common.StoreTimer;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Collect keyed values into a {@link TreeMap} so that they end up sorted.
+ * @param <K> type of key
+ * @param <V> type of value
+ */
+@API(API.Status.EXPERIMENTAL)
+public class MemorySorter<K, V> {
+    @Nonnull
+    private final NavigableMap<K, V> map;
+    @Nonnull
+    private final MemorySortAdapter<K, V> adapter;
+    @Nullable
+    private final StoreTimer timer;
+
+    private LoadResult loadResult;
+
+    public MemorySorter(@Nonnull MemorySortAdapter<K, V> adapter, @Nullable StoreTimer timer) {
+        this.adapter = adapter;
+        this.map = new TreeMap<>(adapter);
+        this.timer = timer;
+    }
+
+    @Nonnull
+    public NavigableMap<K, V> getMap() {
+        return map;
+    }
+
+    @Nonnull
+    public MemorySortAdapter<K, V> getAdapter() {
+        return adapter;
+    }
+
+    public void addKeyValue(K key, V value) {
+        map.put(key, value);
+    }
+
+    public void addValue(V value) {
+        addKeyValue(adapter.generateKey(value), value);
+    }
+
+    /** How the {@code sizeLimit} is interpreted for {@link #load}. */
```

Comment:
Both of those cases are pretty much as described, yes. But I think the bigger context is an answer to the first sentence, which is more or less that it has nothing to do with the semantics of a query because this isn't exposed directly as something the query controls.

In the case of discard, you would get at the second 1000 records by setting a minimum key for records that get loaded so that they are greater than those done the first time. That is done by `MemorySortContinuation`.

In the case of stop, you have to put the first 1000 someplace and later merge that someplace with the next 1000. Which is how the file sorter uses the memory sorter: to do in-memory tree sorting saved to files and then merge sort of those files. (This isn't a particularly innovative approach; that's how sorting works in many SQL databases.)